Add conflict resolution

dsplaisted opened the pull request:
Fixes #974.  Currently work in progress, see the issue for what's done and what's left to do.

EDIT: This should now be ready for review.

dsplaisted committed:
Add conflict resolution and trimming tasks

Copied from dotnet/standard repo, commit a6be13f4bfa9dda899f41dd0983347f2436c084d

dsplaisted committed:
Add conflict resolution tasks to Microsoft.NET.Build.Tasks project

dsplaisted committed:
Update namespace for conflict resolution

dsplaisted committed:
Update license header from conflict resolution code to match other files

dsplaisted committed:
Add conflict resolution .targets file to SDK

dsplaisted committed:
Add tests for conflict resolution scenarios

dsplaisted committed:
Add test for conflict resolution when targeting .NET Framework, update other conflict resolution tests to cover more packages

dsplaisted committed:
Create IConflictItem interface to make ConflictResolver more testable

dsplaisted committed:
Fix incorrect logic (probably copy/paste error) in conflict resolver item type logic

dsplaisted committed:
Add unit tests for conflict resolver

dsplaisted committed:
Don't report a conflict between two items that don't resolve to a file on disk

This avoids conflicts being reported in log files when the same FrameworkAssembly reference is declared in multiple NuGet packages

dsplaisted committed:
Make conflict resolution strings localizable

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/ConflictResolution/RemoveDepsFileConflicts.cs:
```diff
@@ -0,0 +1,156 @@
+
+namespace Microsoft.NET.Build.Tasks.ConflictResolution
+{
+    public partial class RemoveDepsFileConflicts : Task
```
I'm not sure if this was in your plan or not, but instead of generating everything into the .deps.json file, and then removing the conflicts in a separate task, we should just be feeding the Conflicts into the original GenerateDepsFile task.  Then we can generate the whole thing in one shot.

It makes the logic simpler, and more efficient.

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.Packaging.Tools.targets:
```diff
@@ -0,0 +1,133 @@
+<!--
+***********************************************************************************************
+Microsoft.Packaging.Tools.targets
```
Is this the best file name for this targets file?

ericstj replied:
No, in fact it must be changed.

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.Packaging.Tools.targets:
```diff
@@ -0,0 +1,133 @@
+      </PropertyGroup>
+    </When>
+    <Otherwise>
+      <!-- NuGet 2, run before targets that consume references -->
```
I don't think we need to handle NuGet 2 or 3.  We will only ever support the NuGet we are using.

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.Packaging.Tools.targets:
```diff
@@ -0,0 +1,133 @@
+    <!-- Allow completely disabling the conflict resolution targets-->
+    <When Condition="'$(DisableHandlePackageFileConflicts)' == 'true'" />
+    <!-- Condition here is a hack until https://github.com/dotnet/sdk/issues/534 is fixed -->
+    <When Condition="'$(TargetFramework)' != '' or '$(TargetFrameworks)' != ''">
```
We shouldn't need this condition/hack.  We obviously know that the project is using this SDK, because this file is being loaded from our SDK.

ericstj replied:
The only thing we need is a good property to understand when the SDK is providing this so that we can use the version from the standard repo in that case (nuget2/nuget3).

eerhardt commented on src/Tasks/Microsoft.Packaging.Tools/docs/trimming.md:
```diff
@@ -0,0 +1,148 @@
+#Dependency Trimming
```
It was my understanding that trimming was not a 2.0 feature.  Should this be removed?

eerhardt replied:
If we are leaving it in, we should have tests for it.

ericstj replied:
This should not be brought up.  @dsplaisted as we mentioned before the trimming stuff should be left in standard repo.  Just make sure you are able to support the same sort of hook it requires, which is the same as conflict resolution: items to exclude from copy to output by build and publish and exclude from deps file.

eerhardt commented on src/Tasks/Microsoft.Packaging.Tools/pkg/Microsoft.Packaging.Tools.pkgproj:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
```
We shouldn't need this file nor the .builds file.

All these targets and tasks should just come with our SDK just like any other targets and task.

ericstj replied:
Nor the CSProj.  I'd expect this all to be added to your existing project.

eerhardt commented on src/Tasks/Microsoft.Packaging.Tools/tasks/Trimming/Microsoft.NET.Build.Tasks/LockFileCache.cs:
```diff
@@ -0,0 +1,62 @@
+
+namespace Microsoft.NET.Build.Tasks
+{
+    internal class LockFileCache
```
We already have this file in our SDK.  This can be removed.

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.Packaging.Tools.targets:
```diff
@@ -0,0 +1,133 @@
+                              TargetFramework="$(TargetFrameworkMoniker)"
+                              RuntimeIdentifier="$(RuntimeIdentifier)"
+                              PlatformLibraryName="$(MicrosoftNETPlatformLibrary)"
+                              PrivateAssetsPackageReferences="@(PrivateAssetsPackageReference)">
```
This needs the change I made here: https://github.com/dotnet/standard/pull/283/files

dsplaisted committed:
Fix test failures: Don't resolve publish assemblies for a RID, and don't do it if assets file doesn't exist

dsplaisted committed:
Skip test asset path length check on non-Windows

eerhardt commented on test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAHelloWorldProject.cs:
```diff
@@ -162,6 +164,123 @@ public static void Main()
+        void Conflicts_are_resolved_when_publishing(bool selfContained, [CallerMemberName] string callingMethod = "")
+        {
+            var targetFramework = "netcoreapp2.0";
+            var rid = selfContained ? EnvironmentInfo.GetCompatibleRid(targetFramework) : null;
```
It would be good to also test "low fat apps" as well.  See https://github.com/dotnet/sdk/pull/1053.

dsplaisted replied:
I've tried adding a test for this, and it's getting a bunch of DLLs in the output that shouldn't be copied there.  I think it's because the version of the Microsoft.NETCore.App package that we're using (2.0.0-beta-001834-00) has this logic in it:

```xml
  <ItemGroup Condition="'$(RuntimeIdentifier)' == ''">
    <PackageConflictPlatformManifests Include="$(MSBuildThisFileDirectory)Microsoft.NETCore.App.PlatformManifest.txt" />
  </ItemGroup>
```

So the platform manifest isn't getting picked up for rid-specific shared framework apps.  Has this been updated in newer versions of the package?  What version would have a fix?

@eerhardt @ericstj @weshaggard

eerhardt replied:
I just made a fix for that. https://github.com/dotnet/core-setup/pull/1936

That fix is in `2.0.0-preview1-001913-00`.

eerhardt replied:
That build is working its way into the CLI now.  @dotnet-bot should be updating https://github.com/dotnet/cli/pull/6238 with the `1913` version soon.

eerhardt replied:
@dsplaisted - there is a new CLI build that contains this fix:

`2.0.0-preview1-005722`

ericstj commented on src/Tasks/Microsoft.Packaging.Tools/tasks/targets/Microsoft.Packaging.Tools.Trimming.targets:
```diff
@@ -0,0 +1,80 @@
+﻿<?xml version="1.0" encoding="utf-8" ?>
```
Do not copy this...

ericstj commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.Packaging.Tools.targets:
```diff
@@ -0,0 +1,141 @@
+  </PropertyGroup>
+  <Target Name="RemovePublishDepsFileConflicts" AfterTargets="$(RemovePublishDepsFileConflictsAfter)"
+          Inputs="$(PublishDepsFilePath)" Outputs="$(_RemovePublishDepsFileConflictsSemaphore)">
+    <RemoveDepsFileConflicts DepsFilePath="$(PublishDepsFilePath)" Conflicts="@(_ConflictPackageFiles);@(_PublishConflictPackageFiles)" />
```
We need to kill this target and task and add the proper support into to deps file generation to provide a list of things to omit.

dsplaisted committed:
Add test for conflict resolution for RID specific shared framework app

dsplaisted committed:
Simplify conflict resolution targets since we don't need to support down-level versions of NuGet

dsplaisted committed:
Pass conflicts to remove to deps file generation task instead of reading and rewriting it afterwards

dsplaisted committed:
Rename Conflicts property of GenerateDepsFile task to FilesToSkip

dsplaisted committed:
Rename conflict resolution targets file

dsplaisted committed:
Update to new version of CLI, enable RID specific shared framework conflict test, use bundled .NET Core app 2.0 version in tests

dsplaisted committed:
Delete trimming and other files that were copied over from netstandard repo

dsplaisted commented:
@eerhardt @ericstj Thanks for the feedback so far.  You were looking at a work-in-progress PR, I think I've addressed all the feedback and the PR is ready for a real review now.

dsplaisted committed:
Disable tests which depend on bundled versions of .NET Core / .NET Standard when running on full framework

dsplaisted committed:
Merge remote-tracking branch 'upstream/master' into resolve-conflicts-974

# Conflicts:
#	test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopLibrary.cs

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/ConflictResolution/NuGetUtilities.cs:
```diff
@@ -0,0 +1,56 @@
+
+namespace Microsoft.NET.Build.Tasks.ConflictResolution
+{
+    static partial class NuGetUtilities
```
Can this just go in [https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/NuGetUtils.cs](https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/NuGetUtils.cs) so we don't have 2?

eerhardt replied:
Does this need to be `partial`? There is no other partial file.

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/ConflictResolution/NuGetUtilities.cs:
```diff
@@ -0,0 +1,56 @@
+            packageSubPath = null;
+            try
+            {
+                // this method is just a temporary heuristic until we get metadata added to items created by the .NETCore SDK
```
Can we either log an issue for this, or implement it?  We ARE the .NET Core SDK.

dsplaisted replied:
Filed #1091

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenAConflictResolver.cs:
```diff
@@ -0,0 +1,435 @@
+{
+    public class GivenAConflictResolver
+    {
+        //      - Different keys (same metadata) should not result in a conflict
```
In general the approach we've been taking is that the test method name should describe the test, and not have a comment that is duplicate and can get out of date.

dsplaisted replied:
These comments were from a list of scenarios to cover I wrote before writing the tests.  I think the test names are descriptive enough so I can just delete the comments.

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/ConflictResolution/ReferenceComparer.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace Microsoft.NET.Build.Tasks.ConflictResolution
+{
+    internal class ReferenceComparer<T> : IEqualityComparer, IEqualityComparer<T>
```
I don't see this class used.  Can it be removed?

ericstj replied:
Probably.  At one point I was putting ITaskItems into a dictionary or hashset and wanted reference equality (rather than whatever the backing ITaskItem may have implemented).  I believe all cases that were doing this before now wrap the item in a ConflictItem so this isn't needed.

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/ConflictResolution/PlatformManifestReader.cs:
```diff
@@ -0,0 +1,86 @@
+
+            if (!File.Exists(manifestPath))
+            {
+                string errorMessage = string.Format(CultureInfo.InvariantCulture, Strings.CouldNotLoadPlatformManifest,
```
(suggestion) I find an extension method on string named `FormatInvariantCulture` works well to cut down on duplicating this everywhere.

nguerrera replied:
Localized error messages should be formatted with current culture.

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/ConflictResolution/ItemUtilities.cs:
```diff
@@ -0,0 +1,114 @@
+
+namespace Microsoft.NET.Build.Tasks.ConflictResolution
+{
+    static partial class ItemUtilities
```
Can this be refactored with the existing [https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/ITaskItemExtensions.cs](https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/ITaskItemExtensions.cs)?

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/ConflictResolution/ConflictItem.cs:
```diff
@@ -0,0 +1,199 @@
+
+        public ConflictItemType ItemType { get; }
+
+        private bool? exists;
```
`_exists`?  repeated throughout.

ericstj commented on src/Tasks/Microsoft.NET.Build.Tasks/ConflictResolution/FileUtilities.netstandard.cs:
```diff
@@ -0,0 +1,45 @@
+{
+    static partial class FileUtilities
+    {
+        private static Version GetAssemblyVersion(string sourcePath)
```
Depending on what you're targeting in your task DLL you may be able to remove this cross-compilation.  AssemblyName.GetAssemblyName was added back in netstandard2.0.

dsplaisted replied:
Filed #1093

ericstj commented on src/Tasks/Microsoft.NET.Build.Tasks/ConflictResolution/ILog.cs:
```diff
@@ -0,0 +1,85 @@
+    }
+
+
+    public interface ILog
```
Feel free to replace this with whatever logging abstraction is appropriate for your task DLL.

ericstj replied:
Same with the MSBuildLog.cs.

dsplaisted replied:
We don't have one that's easily mockable for unit testing, so I left it as was.

eerhardt replied:
These types don't need to be public, do they?

---
In reply to: [110278890](https://github.com/dotnet/sdk/pull/1052#discussion_r110278890) [](ancestors = 110278890)

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/ConflictResolution/HandlePackageFileConflicts.cs:
```diff
@@ -0,0 +1,183 @@
+
+namespace Microsoft.NET.Build.Tasks.ConflictResolution
+{
+    public partial class HandlePackageFileConflicts : Task
```
should derive from `TaskBase`.

does this need `partial`?

ericstj commented on src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs:
```diff
@@ -103,6 +113,11 @@ protected override void ExecuteCore()
 
+            if (compileFilesToSkip.Any() || runtimeFilesToSkip.Any())
+            {
+                dependencyContext = TrimFilesToSkip(dependencyContext);
```
Is there a way to do this by following the above pattern?  EG: .WithoutFilesToSkip(filesToSkip)

eerhardt replied:
Yes, in general the `GenerateDepsFile` task should be pretty implementation free.  It just turns all the inputs into whatever `DependencyContextBuilder` needs.  And then it writes the file.

Please put this logic into `DependencyContextBuilder`.

---
In reply to: [110251089](https://github.com/dotnet/sdk/pull/1052#discussion_r110251089) [](ancestors = 110251089)

dsplaisted replied:
Sure, it would be possible.  I think it's probably not worth it as there's risk of changing the behavior while refactoring the code, plus the naming is potentially confusing (should it be `WithFilesToSkip` or `WithoutFilesToSkip`?).

ericstj replied:
Once you have some more tests to reduce the risk you should do this.  I suspect the current implementation is very inefficient since it reconstructs the entire dependency context object graph.  File an issue?

ericstj commented on src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj:
```diff
@@ -35,6 +35,10 @@
       <Version>$(MsBuildPackagesVersion)</Version>
       <ExcludeAssets>Runtime</ExcludeAssets>
     </PackageReference>
+    <PackageReference Include="System.Reflection.Metadata" Version="$(SystemReflectionMetadataVersion)" />
```
You don't need this on desktop.

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/ConflictResolution/HandlePackageFileConflicts.cs:
```diff
@@ -0,0 +1,183 @@
+{
+    public partial class HandlePackageFileConflicts : Task
+    {
+        HashSet<ITaskItem> referenceConflicts = new HashSet<ITaskItem>();
```
`private`?  repeated throughout.

ericstj commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.ConflictResolution.targets:
```diff
@@ -0,0 +1,87 @@
+    <ResolvePublishAssemblies ProjectPath="$(MSBuildProjectFullPath)"
+                              AssetsFilePath="$(ProjectAssetsFile)"
+                              TargetFramework="$(TargetFrameworkMoniker)"
+                              RuntimeIdentifier=""
```
Why aren't you seeting RuntimeIdentifier?

dsplaisted replied:
See the comment beforehand: Note that the RuntimeIdentifier is not passed to this task, as we want to process the rid-less assets which will be passed to the compiler.

ericstj replied:
That's not what I created this for, I'm not sure why you changed it.  All the compiler assets are already populated in the reference items.  I created this to get the runtime assets that were never raised to refrencecopylocalpaths because they aren't copied during the build...  https://github.com/dotnet/standard/commit/31f0a1329bf18f54ec987a4735147ebb5dd01314

This is definitely **not** getting compile assets, look at the output: AssembliesToPublish are **runtime** assets.

ericstj replied:
You at least need to fix up the comment here.  Whether or not it requires the RID really depends on the behavior of `build` for a standalone app.  If that's expected to run from the shared framework and it uses the RID-less target for generating the deps file then you can omit the RID here too, I don't think that's the case.  If instead that's expected to run from RID-specific packages same as it would on publish, just without copying, then you must specify the RID here.

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/ConflictResolution/ConflictResolver.cs:
```diff
@@ -0,0 +1,260 @@
+
+
+
+namespace Microsoft.NET.Build.Tasks.ConflictResolution
```
(nit) a bunch of extra newlines.

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/ConflictResolution/ConflictResolver.cs:
```diff
@@ -0,0 +1,260 @@
+            this.packageRank = packageRank;
+        }
+
+        public void ResolveConflicts(IEnumerable<TConflictItem> conflictItems, Func<TConflictItem, string> getItemKey, Action<TConflictItem> foundConflict, bool commitWinner = true,
```
@livarcocc would complain about the length of this line.  :)

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/ConflictResolution/FileUtilities.cs:
```diff
@@ -0,0 +1,34 @@
+            return null;
+        }
+
+        static readonly HashSet<string> s_assemblyExtensions = new HashSet<string>(new[] { ".dll", ".exe", ".winmd" }, StringComparer.OrdinalIgnoreCase);
```
Do we support `winmd`?

ericstj replied:
You ought to.  Don't go out of your way to prevent it.  I'm sure there's a time in the near future where you'll need to make a concerted effort to ensure it works.

dsplaisted replied:
Probably not currently but I imagine we will eventually.

ericstj commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.ConflictResolution.targets:
```diff
@@ -0,0 +1,87 @@
+          DependsOnTargets="ComputePrivateAssetsPackageReferences;
+                            _DefaultMicrosoftNETPlatformLibrary"
+          Condition=" '$(DesignTimeBuild)' != 'true' Or Exists('$(ProjectAssetsFile)')">
+    <!-- Essentially a copy of the SDKs RunResolvePublishAssemblies.
```
You may be able to remove this target if you can somehow make the SDK targets provide the right data.

The purpose of this target is to get all the runtime assemblies as items so that they can be trimmed from the deps file.  When I first implemented this, there wasn't a good item I could use that represented the runtime items at build time but that may have changed or I may not have seen it.

Runtime items are read for publish, since publish needs to do the copy.

I was thinking that if the entire deps content flowed through msbuild this target could certainly go away since the runtime items which eventually get written to the deps file for build would be present.  Barring that it seems like the SDK could at least populate a similar item.  That would really help to get rid of this hack.

eerhardt replied:
Note here that the deps file that is produced for "build" is different than the deps file that is produced for "publish", that is why they are called twice.

One of the main reasons for this is because of the push to make "self-contained" a publish time decision.
Another reason is because `PrivateAssets=All` assets are in the deps file for "build", but they are excluded on "publish".

---
In reply to: [110252304](https://github.com/dotnet/sdk/pull/1052#discussion_r110252304) [](ancestors = 110252304)

ericstj replied:
@eerhardt I'm not talking about the deps file here.  This target has nothing to do with that.

All this target is doing is getting the set of runtime assets that "will be loaded from packages" for build.  It's totally reasonable to have that item populated during build and I'm pretty sure there's already a target that does it.

Moreover, ResolvePublishAssemblies shouldn't even exist.  That should just be a filtered set of the items already raised from the lock file.  I'm fine if that work is put off a bit, but it seems like we could at least run the one target (RunResolvePublishAssemblies) during build, and just make sure that we didn't consume its output (by copying) in the build target closure.

ericstj replied:
I think this still needs to be addressed.  You can file an issue to address in a follow up if you like.  This entire target exists because I implemented this stuff outside the SDK and didn't have enough confidence in what I could take a dependency on.

ericstj commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.ConflictResolution.targets:
```diff
@@ -0,0 +1,87 @@
+    </ResolvePublishAssemblies>
+  </Target>
+
+  <UsingTask TaskName="HandlePackageFileConflicts" AssemblyFile="$(MicrosoftNETBuildTasksAssembly)" />
```
Don't you have some common file for UsingTasks?

dsplaisted replied:
No, not really.  They're scattered throughout different targets files.

ericstj replied:
OK

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj:
```diff
@@ -91,6 +95,7 @@
     <Compile Include="CacheArtifactParser.cs" />
     <Compile Include="EmbedAppNameInHost.cs" />
     <Compile Include="Properties\AssemblyInfo.cs" />
+    <Compile Include="ConflictResolution\**\*.cs" />
```
Should we change this whole thing to use globbing?

dsplaisted replied:
Maybe, but it's not urgent and doesn't need to be part of this change.

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs:
```diff
@@ -8,6 +8,8 @@
 using Microsoft.Extensions.DependencyModel;
 using Newtonsoft.Json;
 using NuGet.ProjectModel;
+using System;
```
(nit) sort usings.

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs:
```diff
@@ -64,8 +69,13 @@ public ITaskItem[] FilesWritten
             get { return _filesWritten.ToArray(); }
         }
 
+        Dictionary<string, HashSet<string>> compileFilesToSkip = new Dictionary<string, HashSet<string>>(StringComparer.OrdinalIgnoreCase);
```
`private`

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs:
```diff
@@ -111,5 +126,115 @@ protected override void ExecuteCore()
+                HashSet<string> filesToSkip;
+                if (runtimeFilesToSkip.TryGetValue(runtimeLibrary.Name, out filesToSkip))
+                {
+                    yield return new RuntimeLibrary(runtimeLibrary.Type,
```
I'm wondering if we should not emit the library if there are no assets in it.  @ericstj @pakrym, what do you think?

That way we don't have a bunch of empty "System.XXX" libraries when they have all been conflict resolved out.

pakrym replied:
I'm not sure, if app referenced a library I think it would be consistent if it was reflected in `.deps` even if `.dll` is coming from the other one.

eerhardt replied:
I can go either way.

But the reason I brought it up is because if NuGet was doing this conflict resolution, you wouldn't see this library at all in the `project.assets.json` file, thus it wouldn't be in the deps file.  However, since NuGet can't do this resolution (because we are collapsing the System.XXX libraries into Microsoft.NETCore.App), these old System.XXX v4 libraries are showing up in the deps file.

pakrym replied:
Let's trim then, I can't come up with a good reason why they are needed and smaller deps file is better.

eerhardt replied:
@dsplaisted - I'm totally fine if you don't make this change in this PR and log a separate issue for this.

dsplaisted replied:
Filed #1109

wli3 replied:
@eerhardt I am working on removing these no asset libraries. However, I am not sure if I should trim these libraries solely on no assets. If there is **no assets** but these is **dependencies**, should I trim it as well?

I am afraid of if there is a "dependencies" that has a asset, removing the parent node will remove the assets that is actually used.

Could you give me pointer to the code to find out? If pull in assetB or not is decided before the code above, I can safely remove library A

![image](https://user-images.githubusercontent.com/6993335/29580659-9ee3d020-872b-11e7-97c6-6f8d38462273.png)

eerhardt replied:
@wli3 - you are right, you can't remove a library based solely on "no assets".  You'd have to take into account its dependencies as well.

> Could you give me pointer to the code to find out?

A pointer to what code?  There are a few consumers of the deps.json file, most notably is the MVC code:

https://github.com/aspnet/Mvc/blob/c8a1bb69142792d1046044bde7082b4e3ac8ac27/src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs

As you can see, MVC will walk all the dependencies looking for assemblies that eventually depend on an MVC assembly.  So yes, if a library has a dependency to another library that eventually has an asset, all the libraries in the dependency chain shouldn't be trimmed.

eerhardt commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToReferenceAnAssembly.cs:
```diff
@@ -26,6 +26,13 @@ public class GivenThatWeWantToReferenceAnAssembly : SdkTest
             string referencerTarget,
             string dependencyTarget)
         {
+            if (UsingFullFrameworkMSBuild)
```
I don't understand why we are making this change here.  It seems irrelevant to your current feature work and shouldn't be modified.

dsplaisted replied:
Previously our .NET Core 2 tests were specifying a RuntimeFrameworkVersion.  I didn't want to do that for the tests I added, so I updated to a new version of the CLI where you get the bundled version of .NET Core automatically.  However, the version of full Framework MSBuild on the CLI doesn't have the props file to set the bundled framework version yet, so I had to disable some tests from running on full framework.

eerhardt replied:
So then why not just leave these tests alone?  Again, changing these tests seems irrelevant to your current work.

dsplaisted replied:
> So then why not just leave these tests alone? Again, changing these tests seems irrelevant to your current work.

Other changes that I made (using the bundled version of .NET Core instead of explicitly setting RuntimeFrameworkVersion) broke these tests on full Framework.  I couldn't just leave them alone because these tests were loading the version of the RuntimeFrameworkVersion to use from DependencyVersions.props, which was the incorrect version after I updated to a new .NET CLI.  I could have updated the version in DependencyVersions.props to match the bundled RuntimeFrameworkVersion, but I don't think that's cleaner than just skipping the tests here.

eerhardt commented on test/Microsoft.NET.TestFramework/TestAsset.cs:
```diff
@@ -175,5 +175,66 @@ private bool IsInBinOrObjFolder(string path)
+        <dependency id=""System.Threading.Timer"" version=""4.3.0"" />
+        <dependency id=""System.Xml.ReaderWriter"" version=""4.3.0"" />
+        <dependency id=""System.Xml.XDocument"" version=""4.3.0"" />
+      </group>";
```
Why hard-code xml here just to parse it below?

dsplaisted replied:
It's a copy/paste from the .nuspec.  It seems less error prone (and easier) to just leave it in that form.

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs:
```diff
@@ -50,6 +52,9 @@ public class GenerateDepsFile : TaskBase
         public ITaskItem[] ReferenceSatellitePaths { get; set; }
 
+        [Required]
+        public ITaskItem[] FilesToSkip { get; set; }
```
We should probably have a test that verifies the .deps.json file is trimmed correctly.  I don't see one.  It could probably go into Microsfot.NET.Build.Tests

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.ConflictResolution.targets:
```diff
@@ -0,0 +1,87 @@
+                              TargetFramework="$(TargetFrameworkMoniker)"
+                              RuntimeIdentifier=""
+                              PlatformLibraryName="$(MicrosoftNETPlatformLibrary)"
+                              PrivateAssetsPackageReferences="@(PrivateAssetsPackageReference)">
```
This should pass in $(IsSelfContained)

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.ConflictResolution.targets:
```diff
@@ -0,0 +1,87 @@
+    <HandlePackageFileConflictsDependsOn Condition="'$(SelfContained)' != 'true'">_GetLockFileAssemblies</HandlePackageFileConflictsDependsOn>
+  </PropertyGroup>
+  
+  <Target Name="HandlePackageFileConflicts" AfterTargets="ResolvePackageDependenciesForBuild" DependsOnTargets="$(HandlePackageFileConflictsDependsOn)">
```
Do we want these "Conflicts" Targets to be part of our "public API"?

eerhardt replied:
If not, we should prefix the name with "_".

---
In reply to: [110262316](https://github.com/dotnet/sdk/pull/1052#discussion_r110262316) [](ancestors = 110262316)

dsplaisted replied:
To me it seems to be on the same level as other targets (eg for generating the deps files) that we define.

"Public API" or not I think it's OK to change these if we need to- ie I don't think we need to guarantee compatibility across versions of the SDK for targets that plug into our SDK.

eerhardt replied:
1. I think these "handle conflicts" tasks are implementation details, and shouldn't be part of our public API.
2. Generating the deps file isn't an implementation detail.  The deps file is a user artifact.
3. I think we need to have an "all up" extension story on what we think we can and can't change across versions of the SDK.  The current approach has been to prefix "private" things with `_`. If that is changing here, we should document it somehow to let people know what our extension story is.

ericstj replied:
I do think folks will depend on doing stuff after these, so we have to imagine an extension point for "where do I sequence a target to get a filtered list of reference/referencecopylocalpaths/deps-file-items.

I agree with @eerhardt that these should not be public.

They should also not occupying the "good spot" for sequencing.  What I mean here is that folks should easily be able to run after them without having ambiguous ordering with these targets themselves.

The names/sequencing they have now are due to me implementing them outside the SDK.  Inside the SDK I think they should be rolled into other target sequences.

Here's a straw man:
1. As a dev extending the SDK I can sequence a target AfterTargets="ResolveReferences" and I am garunteed to get a complete and filtered list of Reference and ReferencePath items.
2. As a dev extending the SDK I can sequence a target AfterTargets="ResolvePackageDependenciesForBuild" and I am garunteed to get a complete and filtered list of items used to construct the build-time deps file.
3. As a dev extending the SDK I can sequence a target AfterTargets="RunResolvePublishAssemblies" and I am garunteed to get a complete and filtered list of ResolvedAssembliesToPublish items (not the case today).
4. As a dev extending the SDK I can sequence a target AfterTargets="ComputeFilesToPublish" and I am garunteed to get a complete and filtered list of ResolvedFileToPublish items, as well as a complete and filtered set of items that will be used to construct the deps file.

Perhaps 1 & 2 and 3 & 4 can be collapsed so long as the items output don't have a relationship (eg one copied to the other).

Moreover, we should have a property that folks can append Targets to so that folks that want to hook this sequence can insert themselves into a deterministic sequence and follow a pattern with common inputs/outputs (a specific set of items) that they can operate on and garuntee those items are the only thing they need to care about updating.

dsplaisted committed:
Consolidate NuGetUtils and NuGetUtilities classes

dsplaisted committed:
Consolidate ITaskItemExtensions and ItemUtilities

dsplaisted committed:
Apply code review feedback

dsplaisted committed:
Disable conflict resolution logic from Microsoft.Packaging.Tools package

dsplaisted committed:
Add tests for deps.json conflict removal

dsplaisted commented:
@eerhardt @ericstj I've addressed the feedback, and added some tests for the deps.json

dsplaisted committed:
Disable test on full framework until #1077 is fixed

dsplaisted committed:
Remove out of date comments about how to disable conflict resolution

dsplaisted committed:
Add tests for running app from output folder, fix issues with conflict resolution when RID is specified

dsplaisted commented:
I've added tests for running an app from the output folder.  As @ericstj suspected, running from the output folder when a RID was specified and there were conflicts on runtime items was broken.

I resolved this by switching from using the `ResolvePublishAssemblies` task in `_GetLockFileAssemblies` to depending on `ComputePrivateAssetsPackageReferences`, which gets the files that will be copied to the output folder if `CopyLocalLockFileAssemblies` is set.  This uses the items from the assets file that flow through MSBuild like we have discussed.  I don't know that the resulting items will be exactly the same as would come from `ResolvePublishAssemblies`, but I think that if there are differences it would be that the publish resolver excludes publishing certain files, which we may not want to exclude from conflict resolution anyway.

dsplaisted commented:
@dotnet-bot test Windows_NT_FullFramework Debug

eerhardt commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.props:
```diff
@@ -13,6 +13,9 @@ Copyright (c) .NET Foundation. All rights reserved.
   <PropertyGroup>
     <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
+
+    <!-- This property disables the conflict resolution logic from the Microsoft.Packaging.Tools package, which is superceded by the logic here in the SDK -->
```
I assume the Packaging.Tools package will change to not be enabled for SDK projects, and then this property can just go away.  Do we have a bug logged to track that?

dsplaisted replied:
I've filed https://github.com/dotnet/standard/issues/298

ericstj commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.ConflictResolution.targets:
```diff
@@ -0,0 +1,68 @@
+
+  <Target Name="_GetLockFileAssemblies"
+          DependsOnTargets="_ComputeLockFileCopyLocal">
+    <!-- We need to find all the files that will be loaded from deps for conflict resolution.
```
This looks much better.  Any reason you didn't use @(AllCopyLocalItems)?  Those look like they'll have the NuGetPackageId metadata which HandlePackageFileConflicts depends on.

ericstj replied:
You can probably eliminate this as a separate target entirely and just have HandlePackageFileConflicts depend on _ComputeLockFileCopyLocal.

dsplaisted committed:
Make conflict resolution targets private

dsplaisted committed:
Don't use Path metadata for destination path in conflict resolution

The AllCopyLocalItems has the Path metadata set to the path of the file within the package, not the target path it will be copied to

dsplaisted committed:
Rename HandlePackageFileConflicts task to ResolvePackageFileConflicts to avoid conflicts with the version from Microsoft.Packaging.Tools

dsplaisted committed:
Merge remote-tracking branch 'upstream/master' into resolve-conflicts-974

dsplaisted committed:
Update test merged from master to work with bundled .NET Core version
